apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/ValueDeserializer.cs

```diff
@@ -0,0 +1,231 @@
+                            ?? NoConversion();
+                    case Value.ValueTypeOneofCase.MapValue:
+                        return DeserializeMap(db, value.MapValue.Fields, targetType);
+                    case ReferenceValue:
+                        return ConvertDefault(v => db.GetDocumentReferenceFromResourceName(v.ReferenceValue))
+                            ?? NoConversion();
+                    default:
+                        return NoConversion();
+                }
+            }
+
+            // Converts the value using the converter if the target type is either object (so general)
+            // or T, or T?.
+            object ConvertDefault<T>(Func<Value, T> conversion) =>
+                targetType == typeof(T) || targetType == Nullable.GetUnderlyingType(typeof(T)) || targetType == typeof(object)
+                ? (object) conversion(value) : null;
+
+            // Converts the value using the converter only if the target type is exactly T or T?.
+            object ConvertSpecific<T>(Func<Value, T> conversion) =>
+                targetType == typeof(T) || targetType == Nullable.GetUnderlyingType(typeof(T))
+                ? (object) conversion(value) : null;
+
+            // Always throws - useful at the end of a chain of attempts.
+            object NoConversion() =>
+                throw new ArgumentException($"Unable to convert value type {value.ValueTypeCase} to {targetType}");
+        }
+
+        private object DeserializeMap(FirestoreDb db, IDictionary<string, Value> values, BclType targetType)
+        {
+            // TODO: Handle dictionaries where TValue isn't object
+            if (targetType == typeof(object) || targetType == typeof(IDictionary<string, object>))
+            {
+                targetType = _defaultMapType;
+            }
+            if (targetType == typeof(Dictionary<string, object>))
+            {
+                var ret = new Dictionary<string, object>();
+                foreach (var pair in values)
+                {
+                    ret[pair.Key] = Deserialize(db, pair.Value, typeof(object));
+                }
+                return ret;
+            }
+            if (targetType == typeof(ExpandoObject))
+            {
+                IDictionary<string, object> ret = new ExpandoObject();
+                foreach (var pair in values)
+                {
+                    ret[pair.Key] = Deserialize(db, pair.Value, typeof(object));
+                }
+                return ret;
+            }
+            if (IsFirestoreAttributedType(targetType))
+            {
+                var ret = Activator.CreateInstance(targetType);
+                // TODO(optimization): We almost certainly want to cache this.
+                var setterMap = new Dictionary<string, PropertyInfo>();
+                foreach (var property in targetType.GetTypeInfo().DeclaredProperties.Where(p => p.CanRead && p.GetSetMethod().IsPublic && !p.GetSetMethod().IsStatic))
```

Comment:
`CanRead` => `CanWrite`?

Reply:
Ooh, missed this comment. Yes. Will fix now.